File: python/3/main.py
def part1(rows):

    right_step, down_step = (3,1)

    trees_encounter = 0

    column = 0
    for row in rows:
        if row[column] == "#":
            trees_encounter += 1
        column = (column + right_step) % len(row)

    return trees_encounter

def part2(rows):

    slopes = [[1,1], [3, 1], [5, 1], [7,1], [1, 2]]
    result = 1

    for right, down in slopes:
        trees = 0
        column = 0
        row = 0

        while row < len(rows):
            if rows[row][column] == "#":
                trees += 1

            row += down
            column = (column + right) % len(rows[0])

        result *= trees

    return result

if __name__ == "__main__":
    # Open the file and readlines
    with open('input') as file:
        # convert lines into array and strip \n
        rows = [row.strip('\n') for row in file.readlines()]
    
    print("The first answer is: {}".format(part1(rows)))
    print("The second answer is: {}".format(part2(rows)))

File: rust/3/src/main.rs
use std::fs::File;
use std::io;
use std::io::{BufRead, BufReader};

fn read_input() -> Result<Vec<String>, io::Error> {
    let filename = "input";
    
    let file = File::open(filename)?;

    let reader = BufReader::new(file);
    let input: Vec<String> = reader.lines().collect::<Result<_, _>>().unwrap();
    
    Ok(input)
}

fn toboggan(forest: &Vec<String>, vx: usize, vy: usize) -> u64 {
    let mut x = 0;
    let mut y = 0;
    let mut tree_count = 0;

    while y < forest.len() {
        if forest[y].chars().nth(x).unwrap() == '#' {
            tree_count += 1;
        }
        x = (x + vx) % forest[0].len();
        y += vy;
    }
    return tree_count
}

fn part1() -> u64{
    let input = read_input();
    if let Err(e) = &input {
        eprintln!("Error occured: {}", e);
        std::process::exit(1);
    }

    let forest = input.unwrap();

    let tree_count = toboggan(&forest, 3, 1);

    return tree_count;
}

fn part2() ->u64{
    let input = read_input();
    if let  Err(e) = &input {
        eprintln!("Error occured: {}", e);
        std::process::exit(1);
    }

    let forest = input.unwrap();

    let tree_results: Vec<u64> = vec![
        toboggan(&forest, 1, 1),
        toboggan(&forest, 3, 1),
        toboggan(&forest, 5, 1),
        toboggan(&forest, 7, 1),
        toboggan(&forest, 1, 2),
    ];

    let tree_count: u64 = tree_results.iter().fold(1, |acc, x| acc * x);

    return tree_count;
}

fn main() {
    let part1_result = part1();
    println!("The first answer is: {}", part1_result);
    let part2_result = part2();
    println!("The second answer is: {}", part2_result);
}
